Changelogs for v1.9.0 (#66)
meilisearch/engine-team: meilisearch-release-changelog/v1.9.0.md

# New features and updates 🔥

## Hybrid search improvements

Since we're focusing on AI innovation, this version introduces multiple improvements and changes related to hybrid search.
More detailed changelog [here](https://meilisearch.notion.site/v1-9-AI-search-changes-e90d6803eca8417aa70a1ac5d0225697#38e6d3adf40e4ef1be14a3c4be39df94).

Done by @dureuill and @irevoire in #4633 and #4649

### ⚠️ Breaking changes of hybrid search usage

- Before v1.9, an empty array in `_vectors.embedder` used to be interpreted as a single embedding of dimension 0 when specifying embeddings in documents. In v1.9 it is now interpreted as 0 embedding. The previous behavior was surprising and not useful.
- From v1.9, by default, `_vectors` will not be returned anymore during a search. This will considerably improve performance. To display it, a new boolean search parameter has been introduced: `retrieveVectors` (default: `false`).

### Improvements

Meilisearch v1.9.0 improves performance when indexing and using hybrid search, avoiding useless operations and optimizing the important ones.

## Filter by score

> TBD

Reviewer: ````suggestion
To filter returned documents by their ranking score, a new `rankingScoreThreshold` parameter has been added to the search and similar routes.

When a `rankingScoreThreshold` is provided, the results of the search/similar request are modified in the following way:

1. No document whose `_rankingScore` is under the `rankingScoreThreshold` is returned
2. Any document encountered during the search that is under the threshold is removed from the set of candidates and won’t count towards the `estimatedTotalHits`, `totalHits` and the facet distribution.

### Examples

request with**out** score threshold:

```json
POST /indexes/movies/search
{
  "q": "Badman dark returns 1",
  "showRankingScore": true,
  "limit": 5
}
```

results:

```json
{
	"hits": [
	    {
	      "title": "Batman the dark knight returns: Part 1",
	      "id": "A",
	      "_rankingScore": 0.93430081300813
	    },
	    {
	      "title": "Batman the dark knight returns: Part 2",
	      "id": "B",
	      "_rankingScore": 0.6685627880184332
	    },
	    {
	      "title": "Badman",
	      "id": "E",
	      "_rankingScore": 0.25
	    },
	    {
	      "title": "Batman Returns",
	      "id": "C",
	      "_rankingScore": 0.11553030303030302
	    },
	    {
	      "title": "Batman",
	      "id": "D",
	      "_rankingScore": 0.11553030303030302
	    }
	],
	"query": "Badman dark returns 1",
	"processingTimeMs": 11,
	"limit": 5,
	"offset": 0,
	"estimatedTotalHits": 62
}
```

request **with** score threshold:

```json
POST /indexes/movies/search
{
  "q": "Badman dark returns 1",
  "showRankingScore": true,
  "limit": 5
  "rankingScoreThreshold": 0.2
}
```

results:

```json
{
	"hits": [
	    {
	      "title": "Batman the dark knight returns: Part 1",
	      "id": "A",
	      "_rankingScore": 0.93430081300813
	    },
	    {
	      "title": "Batman the dark knight returns: Part 2",
	      "id": "B",
	      "_rankingScore": 0.6685627880184332
	    },
	    {
	      "title": "Badman",
	      "id": "E",
	      "_rankingScore": 0.25
	    }
	],
	"query": "Badman dark returns 1",
	"processingTimeMs": 11,
	"limit": 5,
	"offset": 0,
	"estimatedTotalHits": 3
}
```

### Known limitations

⚠️ For performance reasons, if Meilisearch finds `limit` hits above the `rankingScoreThreshold`, then the ranking score of the remaining documents is not evaluated, and so they are not removed from the set of candidates, even if their ranking score would be below the threshold.

As a result, in this configuration the `estimatedTotalHits`, `totalHits` and the facet distribution may be overapproximation of their values.
````

  "offset": 0,
  // Optional, defaults to 20: how many results to display
  "limit": 2,
  // Optional, defaults to null: an additional filter for the returned documents
  "filter": "release_date > 1521763199",
  // Optional, defaults to the default embedder: name of the embedder to use
  // for computing recommendations.
  "embedder": "default",
  // Optional, defaults to null: same as the search query parameter of the same name
  "attributesToRetrieve": [],
  // Both optional, defaults to false: allow displaying the ranking score
  // (resp. detailed ranking score)
  "showRankingScore": false,
  "showRankingScoreDetails": false
}
```

Done by @dureuill in #4647

## `frequency` matching strategy when searching

A `frequency` variant to the `matchingStrategy` search parameter has been added. This favors the least frequent query words when retrieving the documents.

```bash
curl \
 -X POST 'http://localhost:7700/indexes/movies/search' \
 -H 'Content-Type: application/json' \
 --data-binary '{
    "q": "chaval blanc",
    "matchingStrategy": "frequency"
 }'
```

Previous existing values for `matchingStrategy` are `last` and `all` (`last` is the default value).

> Done by @ManyTheFish in TBD

Author: ```suggestion
Done by @ManyTheFish in #4667
```

Meilisearch now limits operations when importing settings by avoiding useless writing operations in its internal database and by reducing disk usage.

Done by @irevoire and @Kerollmops in #4646, #4656 and #4631

## Other improvements

* [Prometheus experimental feature](https://github.com/orgs/meilisearch/discussions/625): Use HTTP path pattern instead of full path in metrics (#4619) @gh2k
* ⚠️ Remove `exportPuffinReport` experimental feature. Use [logs routes](https://github.com/orgs/meilisearch/discussions/721) and [logs modes](https://github.com/orgs/meilisearch/discussions/723) instead (#4655) @Kerollmops

# Fixes 🐞

* When no searchable attributes are declared, all the fields have the same importance instead of being randomly given more importance. More information [here](https://github.com/meilisearch/meilisearch/issues/4639) (#4631) @irevoire
* Fix `searchableAttributes` behavior with nested fields when they were not explicitly defined. More information [here](https://github.com/meilisearch/meilisearch/issues/4639) (#4631) @irevoire
* Fix security issue in dependency: bump Rustls to non-vulnerable versions (#4622) @Kerollmops

# Misc

* Dependencies updates

>   * TBD

Reviewer: ```suggestion
  * Update actix-web 4.5.1 -> 4.6.0 done by @dureuill in https://github.com/meilisearch/meilisearch/pull/4675
```

* CIs and tests
  * Add "precommands" to benchmark (#4624) @dureuill
  * Allow to comment with the results of benchmark invocation (#4651) @dureuill
* Documentation

>   * TBD

Author: ```suggestion
  * Update README.md (#4664) @tpayet
```

🧰 All official Meilisearch integrations (including SDKs, clients, and other tools) are compatible with this Meilisearch release. Integration deployment happens between 4 to 48 hours after a new version becomes available.

<!-- The following lines should NOT be put in the PRE-release changelogs -->
Some SDKs might not include all new features. Consult the project repository for detailed information. Is a feature you need missing from your chosen SDK? Create an issue letting us know you need it, or, for open-source karma points, open a PR implementing it (we'll love you for that ❤️).

# New features and updates 🔥

## Hybrid search updates

This release introduces multiple [hybrid search updates](https://meilisearch.notion.site/v1-9-AI-search-changes-e90d6803eca8417aa70a1ac5d0225697#38e6d3adf40e4ef1be14a3c4be39df94).

Done by @dureuill and @irevoire in #4633 and #4649

### ⚠️ Breaking change: Empty `_vectors.embedder` arrays

Empty `_vectors.embedder` arrays are now interpreted as having no vector embedding.

Before v1.9, Meilisearch interpreted these as a single embedding of dimension 0. This change follows user feedback that the previous behavior was unexpected and unhelpful.

> 

Reviewer: Should we say here that Meilisearch is returning the float in a canonicalized representation rather than the user-provided representation?

> 

Author: Yes I think it's better to say it like this ("returning the float in a canonicalized representation rather than the user-provided representation") with the example of tamo

> 

Author: I updated it!

### ⚠️ Breaking change: `_vectors` field no longer present in search results

> Meilisearch no longer includes `_vectors` in returned search results by default. This will considerably improve performance.

Reviewer: ```suggestion
When the experimental `vectorStore` feature is enabled, Meilisearch no longer includes `_vectors` in returned search results by default. This will considerably improve performance.
```

Use the new `retrieveVectors` search parameter to display the `_vectors` field:

```sh
curl \
  -X POST 'http://localhost:7700/indexes/INDEX_NAME/search' \
  -H 'Content-Type: application/json' \
  --data-binary '{ 
    "q": "SEARCH QUERY",
    "retrieveVectors": true
  }'
```

### Document `_vector` accepts object values

The document `_vector` field now accepts objects in addition to embedding arrays:

> ```json
> {
>   "id": 42,
>   "_vectors": {
>     "default": [0.1, 0.2 ],
>       "text": {
>         "embeddings": [[0.1, 0.2, 0.3], [0.4, 0.5, 0.6]],
>         "userProvided": true
>       },
>       "translation": {
>         "embeddings": [0.1, 0.2, 0.3, 0.4],
>         "userProvided": false
>       }
>     }
> }
> ```

Reviewer: ```suggestion
```json
{
  "id": 42,
  "_vectors": {
    "default": [0.1, 0.2 ],
      "text": {
        "embeddings": [[0.1, 0.2, 0.3], [0.4, 0.5, 0.6]],
        "regenerate": false
      },
      "translation": {
        "embeddings": [0.1, 0.2, 0.3, 0.4],
        "regenerate": true
      }
    }
}
```
```

The `_vector` object must contain two fields: `embeddings` and `userProvided`.

> `embeddings` must be the document's embeddings.

Reviewer: ```suggestion
If present, `embeddings` will replace this document's embeddings.
```

> `userProvided` must be either `true` or `false`. If `userProvided: true`, Meilisearch uses the configured values and does not overwrite them when re-indexing items with an autoembedder. If `userProvided: false`, Meilisearch will use the configured document embedding, but will overwrite it when re-indexing items with an autoembedder.

Reviewer: ```suggestion
`regenerate` must be either `true` or `false`. If `regenerate: true`, Meilisearch will overwrite the document embeddings each time the document is updated in the future. If `regenerate: false`, Meilisearch will keep the last provided or generated embeddings even if the document is updated in the future.
```

> This change allows importing embeddings to autoembedders as a one-shot process, by setting them as `userProvided: false`. This change also ensures embeddings are not regenerated when importing a dump created with Meilisearch v1.9.

Reviewer: ```suggestion
This change allows importing embeddings to autoembedders as a one-shot process, by setting them as `regenerate: true`. This change also ensures embeddings are not regenerated when importing a dump created with Meilisearch v1.9.
```

Done by @dureuill in #4666

## New feature: Get similar documents endpoint

This release introduces a new AI-powered search feature allowing you to send a document to Meilisearch and receive a list of similar documents in return.

Use the `/indexes/{indexUid}/similar` endpoint to query Meilisearch for related documents:

```sh
curl \
  -X POST /indexes/:indexUid/similar
  -H 'Content-Type: application/json' \
  --data-binary '{
    "id": "23",
    "offset": 0,
    "limit": 2,
    "filter": "release_date > 1521763199",
    "embedder": "default",
    "attributesToRetrieve": [],
    "showRankingScore": false,
    "showRankingScoreDetails": false
  }'
```

- `id`: string indicating the document needing similar results, required
- `offset`: number of results to skip when paginating, optional, defaults to `0`
- `limit`: number of results to display, optional, defaults to `20`
- `filter`: string with a filter expression Meilisearch should apply to the results, optional, defaults to `null`
- `embedder`: string indicating the embedder Meilisearch should use to retrieve similar documents, optional, defaults to `"default"`

> - `attributesToRetrieve`: array of strings indicating which fields Meilisearch will include in the response, optional, defaults to `[]`

Reviewer: Yes it should be the same as the [search one](https://www.meilisearch.com/docs/reference/api/search#attributes-to-retrieve), which defaults to `[*]`

-  `showRankingScore`: boolean indicating if results should include ranking score information, optional, defaults to `false`
- `showRankingScoreDetails`: boolean indicating if results should include detailed ranking score information, optional, defaults to `false`

`/indexes/{indexUid}/similar` supports `GET` and `POST` routes. Use URL query parameters to configure your `GET` request, or include your parameters in the request body if using the `POST` route.  Both offer identical functionality.

Done by @dureuill in #4647

## New feature: `frequency` matching strategy

This release adds a new matching strategy, `frequency`. Use it to prioritize results containing the least frequent query terms:

```bash
curl \
 -X POST 'http://localhost:7700/indexes/{index_uid}/search' \
 -H 'Content-Type: application/json' \
 --data-binary '{

>     "q": "chaval blanc",

Author: ```suggestion
    "q": "cheval blanc",
```

    "matchingStrategy": "frequency"
 }'
```

Done by @ManyTheFish in #4667

> ## Set `distinctAttribute` at search
> 
> TBD
> 

Reviewer: ````suggestion
## Set `distinctAttribute` at search time

This release introduces a new search parameter: `distinct` which you can use to specify the distinct attribute at search time:

```bash
curl \
 -X POST 'http://localhost:7700/indexes/{index_uid}/search' \
 -H 'Content-Type: application/json' \
 --data-binary '{
    "q": "kefir le double poney",
    "distinct": "book.isbn"
 }'
\`\`\`

If a distinct attribute is already defined in the settings it'll be ignored in favor of the one defined at search time.

Done by @Kerollmops in #4693

````
I made the code block work, you need 4 `\`` for the suggestion code block

## Other improvements

* Improve language support (#4684) @ManyTheFish @Soham1803 @mosuka @tkhshtsh0917
  * Add new normalizer to normalize œ to oe and æ to ae
  * Fix `chinese-normalization-pinyin` feature flag compilation
* [Prometheus experimental feature](https://github.com/orgs/meilisearch/discussions/625): Use HTTP path pattern instead of full path in metrics (#4619) @gh2k
* ⚠️ Remove `exportPuffinReport` experimental feature. Use [logs routes](https://github.com/orgs/meilisearch/discussions/721) and [logs modes](https://github.com/orgs/meilisearch/discussions/723) instead (#4655) @Kerollmops

# Fixes 🐞

* All fields now have the same impact on relevancy when `searchableAttributes: ["*"]`. Consult the [GitHub issue for a detailed breakdown of these changes](https://github.com/meilisearch/meilisearch/issues/4639) (#4631) @irevoire
* Fix `searchableAttributes` behavior when handling nested fields. Consult the [GitHub issue for more information](https://github.com/meilisearch/meilisearch/issues/4639) (#4631) @irevoire
* Fix security issue in dependency: bump Rustls to non-vulnerable versions (#4622) @Kerollmops

> 

Reviewer: ```suggestion
* Reset other embedding settings when changing the `source` of an embedder. This prevents [misleading error messages when configuring the embedders](https://github.com/meilisearch/meilisearch/issues/4690) (#4649) @dureuill 

```

```bash
curl \
 -X POST 'http://localhost:7700/indexes/{index_uid}/search' \
 -H 'Content-Type: application/json' \
 --data-binary '{
    "q": "kefir le double poney",
    "distinct": "book.isbn"
 }'
```

If a distinct attribute is already defined in the settings it'll be ignored in favor of the one defined at search time.

Done by @Kerollmops in #4693

## Improve indexing speed when updating/adding settings

Meilisearch now limits operations when importing settings by avoiding unnecessary writing operations in its internal database and reducing disk usage.

Additionally, when changing embedding settings, Meilisearch will now only regenerate the embeddings for the embedders whose settings have been modified, instead of for all embedders. When only the `documentTemplate` is modified, embeddings will only be regenerated for documents where this modification leads to a different text to embed.

Done by @irevoire, @Kerollmops, @ManyTheFish and @dureuill in #4646, #4680, #4631 and #4649

## Other improvements

* Speed up filter ANDs operations (#4682) @Kerollmops

> * Improve language support (#4684) @ManyTheFish @Soham1803 @mosuka @tkhshtsh0917

Reviewer: ```suggestion
* Speed up facet distribution (#4713) @Kerollmops 
* Improve language support (#4684) @ManyTheFish @Soham1803 @mosuka @tkhshtsh0917
```

  * Fix `chinese-normalization-pinyin` feature flag compilation
* [Prometheus experimental feature](https://github.com/orgs/meilisearch/discussions/625): Use HTTP path pattern instead of full path in metrics (#4619) @gh2k
* ⚠️ Remove `exportPuffinReport` experimental feature. Use [logs routes](https://github.com/orgs/meilisearch/discussions/721) and [logs modes](https://github.com/orgs/meilisearch/discussions/723) instead (#4655) @Kerollmops

# Fixes 🐞

* All fields now have the same impact on relevancy when `searchableAttributes: ["*"]`. Consult the [GitHub issue for a detailed breakdown of these changes](https://github.com/meilisearch/meilisearch/issues/4639) (#4631) @irevoire
* Fix `searchableAttributes` behavior when handling nested fields. Consult the [GitHub issue for more information](https://github.com/meilisearch/meilisearch/issues/4639) (#4631) @irevoire
* Fix security issue in dependency: bump Rustls to non-vulnerable versions (#4622) @Kerollmops
* Reset other embedding settings when changing the `source` of an embedder. This prevents [misleading error messages when configuring the embedders](https://github.com/meilisearch/meilisearch/issues/4690) (#4649) @dureuill

> 

Reviewer: ```suggestion
* Fix panic in hybrid search when removing all embedders from the DB (https://github.com/meilisearch/meilisearch/pull/4715) @irevoire 

```

# Misc

* Dependencies updates

>   * Update actix-web 4.5.1 -> 4.6.0 (#4675) @dureuill

Reviewer: ```suggestion
  * Update actix-web 4.5.1 -> 4.6.0 (#4675) @dureuill
  * Update mini-dashboard to 2.13 -> 2.14 (#4712) @curquiza 
```

### ⚠️ Breaking change: Empty `_vectors.embedder` arrays

Empty `_vectors.embedder` arrays are now interpreted as having no vector embedding.

Before v1.9, Meilisearch interpreted these as a single embedding of dimension 0. This change follows user feedback that the previous behavior was unexpected and unhelpful.

### ⚠️ Breaking change: `_vectors` field no longer present in search results

When the experimental `vectorStore` feature is enabled, Meilisearch no longer includes `_vectors` in returned search results by default. This will considerably improve performance.

Use the new `retrieveVectors` search parameter to display the `_vectors` field:

```sh
curl \
  -X POST 'http://localhost:7700/indexes/INDEX_NAME/search' \
  -H 'Content-Type: application/json' \
  --data-binary '{
    "q": "SEARCH QUERY",
    "retrieveVectors": true
  }'
```

### ⚠️ Breaking change: `_vectors.embedder` arrays can be altered by Meilisearch

In order to save storage and run faster, Meilisearch is no longer storing your vector "as-is". Meilisearch now returns the float in a canonicalized representation rather than the user-provided representation.

For example, `3` may be represented as `3.0`

> ### Document `_vector` accepts object values

Reviewer 2: ```suggestion
### Document `_vectors` accepts object values
```

> The document `_vector` field now accepts objects in addition to embedding arrays:

Reviewer 3: ```suggestion
### Document `_vectors` accepts object values

The document `_vectors` field now accepts objects in addition to embedding arrays:
```

> The document `_vector` field now accepts objects in addition to embedding arrays:

Reviewer 2: ```suggestion
The document `_vectors` field now accepts objects in addition to embedding arrays:
```

 -H 'Content-Type: application/json' \
 --data-binary '{
    "q": "Badman dark returns 1",
    "showRankingScore": true,
    "limit": 5,
    "rankingScoreThreshold": 0.2
 }'
```

Meilisearch does not return any documents below the configured threshold. Excluded results do not count towards `estimatedTotalHits`, `totalHits`, and facet distribution.

⚠️ For performance reasons, if the number of documents above `rankingScoreThreshold` is higher than `limit`, Meilisearch does not evaluate the ranking score of the remaining documents. Results ranking below the threshold are not immediately removed from the set of candidates. In this case, Meilisearch may overestimate the count of `estimatedTotalHits`, `totalHits` and facet distribution.

Done by @dureuill in #4666

## New feature: Get similar documents endpoint

This release introduces a new AI-powered search feature allowing you to send a document to Meilisearch and receive a list of similar documents in return.

Use the `/indexes/{indexUid}/similar` endpoint to query Meilisearch for related documents:

```sh
curl \
  -X POST /indexes/:indexUid/similar
  -H 'Content-Type: application/json' \
  --data-binary '{

>     "id": "23",

Reviewer 2: What if the ID field name is different, e.g. "productID"?

Author: I guess the primary key is used here. @dureuill

Reviewer: that's an interesting remark from a user perspective.

The parameter name is `id` we always use the `primaryKey` of the document regardless of its name.
Using `productID` here as the parameter name will not work.

Reviewer 2: Thank you for the clarification!

  }'
```

- `id`: string indicating the document needing similar results, required
- `offset`: number of results to skip when paginating, optional, defaults to `0`
- `limit`: number of results to display, optional, defaults to `20`
- `filter`: string with a filter expression Meilisearch should apply to the results, optional, defaults to `null`
- `embedder`: string indicating the embedder Meilisearch should use to retrieve similar documents, optional, defaults to `"default"`
- `attributesToRetrieve`: array of strings indicating which fields Meilisearch will include in the response, optional, defaults to `["*"]`
-  `showRankingScore`: boolean indicating if results should include ranking score information, optional, defaults to `false`
- `showRankingScoreDetails`: boolean indicating if results should include detailed ranking score information, optional, defaults to `false`

`/indexes/{indexUid}/similar` supports `GET` and `POST` routes. Use URL query parameters to configure your `GET` request, or include your parameters in the request body if using the `POST` route.  Both offer identical functionality.

Done by @dureuill in #4647

## New feature: `frequency` matching strategy

> This release adds a new matching strategy, `frequency`. Use it to prioritize results containing the least frequent query terms:

Reviewer 4: No, we prioritize the least frequent words because they are the most discriminating ones.
For instance, the query `Soup of the day`, `of` and `the` are really frequent in documents so they are not really useful to rank documents. We would prefer to focus on `Soup` and `Day`.

> This release adds a new matching strategy, `frequency`. Use it to prioritize results containing the least frequent query terms:

Reviewer 2: Thanks for the clarification @ManyTheFish. 

May I suggest reconsidering the naming of the option so it aligns better with the naming of others? Something like `discriminate` or `infrequent` would seem more logical. 

@guimachiavelli or @CaroFG may have better suggestions than me 😅

> This release adds a new matching strategy, `frequency`. Use it to prioritize results containing the least frequent query terms:

Reviewer: IMO `frequency` does not tell if we're using the most frequent or least frequent. From here, the name sounds apt, and I believe it is a term of art too 🤔

> This release adds a new matching strategy, `frequency`. Use it to prioritize results containing the least frequent query terms:

Reviewer: Wanted to post a longer response, because this is incorrect:

> This naming seems inconsistent with the naming of other options:
> • `all` prioritizes results with all query terms
> • `last` prioritizes results where the last query term is present

Matching strategies don't prioritize results. They define the order in which words are iteratively removed from the query:

- `all`: will remove all words from the query at once, meaning only documents with all words will match.
- `last`: each iteration will remove the last word from the query, meaning that the first iteration requires all words, then the second iteration requires all words minus the last one, etc.
- `frequency`: each iteration will remove the word with the highest frequency from the query, meaning that the first iteration requires all words, then the second iteration requires all words minus the most frequent one, etc.

The name is consistent with this model, which is how the engine handles matching strategies internally.
I think straying from this mental model to go into "prioritizes result" mental models results in a blurry mental model where it is unclear how the matching strategy actually affect the query, and its link with the `words` score details (which displays the number of `maxMatchingWords` and number of remaining `matchingWords`).

In conclusion, I strongly believe the name should remain `frequency`.

* ⚠️ Remove `exportPuffinReport` experimental feature. Use [logs routes](https://github.com/orgs/meilisearch/discussions/721) and [logs modes](https://github.com/orgs/meilisearch/discussions/723) instead (#4655) @Kerollmops

# Fixes 🐞

* All fields now have the same impact on relevancy when `searchableAttributes: ["*"]`. Consult the [GitHub issue for a detailed breakdown of these changes](https://github.com/meilisearch/meilisearch/issues/4639) (#4631) @irevoire
* Fix `searchableAttributes` behavior when handling nested fields. Consult the [GitHub issue for more information](https://github.com/meilisearch/meilisearch/issues/4639) (#4631) @irevoire
* Fix security issue in dependency: bump Rustls to non-vulnerable versions (#4622) @Kerollmops
* Reset other embedding settings when changing the `source` of an embedder. This prevents [misleading error messages when configuring the embedders](https://github.com/meilisearch/meilisearch/issues/4690) (#4649) @dureuill
* Fix panic in hybrid search when removing all embedders from the DB (#4715) @irevoire

> 

Reviewer: ```suggestion
* Hybrid search now respects the `offset` and `limit` parameters when returning keyword results early (#4746) @dureuill

```

Use the `/indexes/{indexUid}/similar` endpoint to query Meilisearch for related documents:

```sh
curl \
  -X POST /indexes/:indexUid/similar
  -H 'Content-Type: application/json' \
  --data-binary '{
    "id": "23",
    "offset": 0,
    "limit": 2,
    "filter": "release_date > 1521763199",
    "embedder": "default",
    "attributesToRetrieve": [],
    "showRankingScore": false,
    "showRankingScoreDetails": false
  }'
```

- `id`: string indicating the document needing similar results, required
- `offset`: number of results to skip when paginating, optional, defaults to `0`
- `limit`: number of results to display, optional, defaults to `20`
- `filter`: string with a filter expression Meilisearch should apply to the results, optional, defaults to `null`
- `embedder`: string indicating the embedder Meilisearch should use to retrieve similar documents, optional, defaults to `"default"`
- `attributesToRetrieve`: array of strings indicating which fields Meilisearch will include in the response, optional, defaults to `["*"]`
-  `showRankingScore`: boolean indicating if results should include ranking score information, optional, defaults to `false`

> - `showRankingScoreDetails`: boolean indicating if results should include detailed ranking score information, optional, defaults to `false`

Reviewer 3: ```suggestion
- `showRankingScoreDetails`: boolean indicating if results should include detailed ranking score information, optional, defaults to `false`
- `rankingScoreThreshold`: Excludes search results with a ranking score lower than the defined number, optional, defaults to `null`.
```

<!-- The following lines should NOT be put in the PRE-release changelogs -->
🧰 All official Meilisearch integrations (including SDKs, clients, and other tools) are compatible with this Meilisearch release. Integration deployment happens between 4 to 48 hours after a new version becomes available.

<!-- The following lines should NOT be put in the PRE-release changelogs -->
Some SDKs might not include all new features. Consult the project repository for detailed information. Is a feature you need missing from your chosen SDK? Create an issue letting us know you need it, or, for open-source karma points, open a PR implementing it (we'll love you for that ❤️).

# New features and updates 🔥

## Hybrid search updates

This release introduces multiple [hybrid search updates](https://meilisearch.notion.site/v1-9-AI-search-changes-e90d6803eca8417aa70a1ac5d0225697#38e6d3adf40e4ef1be14a3c4be39df94).

Done by @dureuill and @irevoire in #4633 and #4649

> ### ⚠️ Breaking change: Empty `_vectors.embedder` arrays

Reviewer 2: Is `_vectors.embedder` equivalent to `_vectors.default.embedder`, or is there an error here?

Reviewer: `_vectors.embedder` talks about the embeddings of a configuration named `embedder`.

You are talking about a configuration named `default` (which we often use in examples).

As a reminder, since user can choose the name of their embedders, you can have `_vectors.arbitrary_name`

Empty `_vectors.embedder` arrays are now interpreted as having no vector embedding.

Before v1.9, Meilisearch interpreted these as a single embedding of dimension 0. This change follows user feedback that the previous behavior was unexpected and unhelpful.

### ⚠️ Breaking change: `_vectors` field no longer present in search results

When the experimental `vectorStore` feature is enabled, Meilisearch no longer includes `_vectors` in returned search results by default. This will considerably improve performance.

Use the new `retrieveVectors` search parameter to display the `_vectors` field:

```sh
curl \
  -X POST 'http://localhost:7700/indexes/INDEX_NAME/search' \
  -H 'Content-Type: application/json' \
  --data-binary '{
    "q": "SEARCH QUERY",
    "retrieveVectors": true
  }'
```

> ### ⚠️ Breaking change: `_vectors.embedder` arrays can be altered by Meilisearch

Reviewer 2: Did we mean `_vectors.embedder.embeddings`?

Reviewer: no, we mean the short-hand form where embeddings appear directly in `_vectors.<embedder_name>`.

As this change is also true for the long form, maybe a clearer way of describing this would be:

```suggestion
### ⚠️ Breaking change: Meilisearch no longer preserves the exact representation of embeddings appearing in `_vectors`
```

I think "not preserving the exact representation" is also more accurate than "alter"

  // for computing recommendations.
  "embedder": "default",
  // Optional, defaults to null: same as the search query parameter of the same name
  "attributesToRetrieve": [],
  // Both optional, defaults to false: allow displaying the ranking score
  // (resp. detailed ranking score)
  "showRankingScore": false,
  "showRankingScoreDetails": false
}
```

Done by @dureuill in #4647

## `frequency` matching strategy when searching

A `frequency` variant to the `matchingStrategy` search parameter has been added. This favors the least frequent query words when retrieving the documents.

```bash
curl \
 -X POST 'http://localhost:7700/indexes/movies/search' \
 -H 'Content-Type: application/json' \
 --data-binary '{
    "q": "chaval blanc",
    "matchingStrategy": "frequency"
 }'
```

Previous existing values for `matchingStrategy` are `last` and `all` (`last` is the default value).

Done by @ManyTheFish in #4667

## Set `distinctAttribute` at search

TBD

## Improve indexing speed when updating/adding settings

Meilisearch now limits operations when importing settings by avoiding useless writing operations in its internal database and by reducing disk usage.

> Done by @irevoire and @Kerollmops in #4646, #4656 and #4631

Author: ```suggestion
Done by @irevoire and @Kerollmops in #4646, ##4680 and #4631
```

> Done by @irevoire and @Kerollmops in #4646, #4680 and #4631

Author: ```suggestion
Done by @irevoire, @Kerollmops and @ManyTheFish in #4646, #4680 and #4631
```

```json
{
  "id": 42,
  "_vectors": {
    "default": [0.1, 0.2 ],
      "text": [[0.1, 0.2, 0.3], [0.4, 0.5, 0.6]],
      "image": null
  }
}
```

```json
{
  "id": 42,
  "_vectors": {
    "default": [0.1, 0.2 ],
      "text": {
        "embeddings": [[0.1, 0.2, 0.3], [0.4, 0.5, 0.6]],
        "userProvided": true
      },
      "translation": {
        "embeddings": [0.1, 0.2, 0.3, 0.4],
        "userProvided": false
      }
    }
}
```

In v1.9, this syntax is extended to allow for the value to be an object in addition to the current embeddings or array of embeddings.

```json
{
  "id": 42,
  "_vectors": {
    "default": [0.1, 0.2 ],
      "text": {
        "embeddings": [[0.1, 0.2, 0.3], [0.4, 0.5, 0.6]],
        "userProvided": true
      },
      "translation": {
        "embeddings": [0.1, 0.2, 0.3, 0.4],
        "userProvided": false
      }
    }
}
```

This change allows **importing** embeddings to autoembedders as a one-shot process, by setting them as `userProvided: false`.

The dump system makes use of this, so with this change embeddings won’t be regenerated when importing a dump created with Meilisearch v1.9.

> Meilisearch v1.9.0 also improves performance when indexing and using hybrid search, avoiding useless operations and optimizing the important ones.

Author: Shouldn't it be `###` and on `##`?

```json
{
	"hits": [
	    {
	      "title": "Batman the dark knight returns: Part 1",
	      "id": "A",
	      "_rankingScore": 0.93430081300813
	    },
	    {
	      "title": "Batman the dark knight returns: Part 2",
	      "id": "B",
	      "_rankingScore": 0.6685627880184332
	    },
	    {
	      "title": "Badman",
	      "id": "E",
	      "_rankingScore": 0.25
	    },
	    {
	      "title": "Batman Returns",
	      "id": "C",
	      "_rankingScore": 0.11553030303030302
	    },
	    {
	      "title": "Batman",
	      "id": "D",
	      "_rankingScore": 0.11553030303030302
	    }
	],
	"query": "Badman dark returns 1",
	"processingTimeMs": 11,
	"limit": 5,
	"offset": 0,
	"estimatedTotalHits": 62
}
```

⚠️ For performance reasons, if Meilisearch finds `limit` hits above the `rankingScoreThreshold`, then the ranking score of the remaining documents is not evaluated, and so they are not removed from the set of candidates, even if their ranking score would be below the threshold. As a result, in this configuration the `estimatedTotalHits`, `totalHits` and the facet distribution may be overapproximation of their values.

Done by @dureuill in #4666

> ## Get similar documents
> 
> To retrieve similar documents in your datasets, two new routes have been introduced
> - `POST /indexes/:indexUid/similar` using parameters in the request body.
> - `GET /indexes/:indexUid/similar`, using query URL parameters.
> 
> ```json5
> POST /indexes/:indexUid/similar
> {
>   // Mandatory: the external id of the target document
>   "id": "23",
>   // Optional, defaults to 0: how many results to skip
>   "offset": 0,
>   // Optional, defaults to 20: how many results to display
>   "limit": 2,
>   // Optional, defaults to null: an additional filter for the returned documents
>   "filter": "release_date > 1521763199",
>   // Optional, defaults to the default embedder: name of the embedder to use
>   // for computing recommendations.
>   "embedder": "default",
>   // Optional, defaults to null: same as the search query parameter of the same name
>   "attributesToRetrieve": [],
>   // Both optional, defaults to false: allow displaying the ranking score
>   // (resp. detailed ranking score)
>   "showRankingScore": false,
>   "showRankingScoreDetails": false
> }
> ```

Reviewer 5: ````suggestion
## New feature: Get similar documents endpoint

This release introduces a new AI-powered search feature allowing you to send a document to Meilisearch and receive a list of similar documents in return.

Use the `/indexes/{indexUid}/similar` endpoint to query Meilisearch for related documents:

```sh
curl \
  -X POST /indexes/:indexUid/similar
  -H 'Content-Type: application/json' \
  --data-binary '{
    "id": "23",
    "offset": 0,
    "limit": 2,
    "filter": "release_date > 1521763199",
    "embedder": "default",
    "attributesToRetrieve": [],
    "showRankingScore": false,
    "showRankingScoreDetails": false
  }'
```

- `id`: string indicating the document needing similar results, required
- `offset`: number of results to skip when paginating, optional, defaults to `0`
- `limit`: number of results to display, optional, defaults to `20`
- `filter`: string with a filter expression Meilisearch should apply to the results, optional, defaults to `null`
- `embedder`: string indicating the embedder Meilisearch should use to retrieve similar documents, optional, defaults to `"default"`
- `attributesToRetrieve`: array of strings indicating which fields Meilisearch will include in the response, optional, defaults to `[]`
-  `showRankingScore`: boolean indicating if results should include ranking score information, optional, defaults to `false`
- `showRankingScoreDetails`: boolean indicating if results should include detailed ranking score information, optional, defaults to `false`

`/indexes/{indexUid}/similar` supports `GET` and `POST` routes. Use URL query parameters to configure your `GET` request, or include your parameters in the request body if using the `POST` route.  Both offer identical functionality.
````

Done by @dureuill in #4647

> ## `frequency` matching strategy when searching
> 
> A `frequency` variant to the `matchingStrategy` search parameter has been added. This favors the least frequent query words when retrieving the documents.
> 
> ```bash
> curl \
>  -X POST 'http://localhost:7700/indexes/movies/search' \
>  -H 'Content-Type: application/json' \
>  --data-binary '{
>     "q": "chaval blanc",
>     "matchingStrategy": "frequency"
>  }'
> ```
> 
> Previous existing values for `matchingStrategy` are `last` and `all` (`last` is the default value).

Reviewer 5: ````suggestion
## New feature: `frequency` matching strategy

This release adds a new matching strategy, `frequency`. Use it to prioritize results containing the least frequent query terms:

```bash
curl \
 -X POST 'http://localhost:7700/indexes/{index_uid}/search' \
 -H 'Content-Type: application/json' \
 --data-binary '{
    "q": "chaval blanc",
    "matchingStrategy": "frequency"
 }'
```
````
